File: app-server/src/terminals/attachProtocol.ts
import { Option, Schema } from "effect"
import type { SessionSize } from "./terminalAdapter.ts"

// The attach wire protocol (ATC-130), stated once for every side of the
// socket: the server bridge (terminalAttach.ts), the CLI client
// (attachClient.ts), and any future client. The contract endpoint
// description (contract.ts `attachTerminal`) documents it in prose; this module
// is the code authority both sides import.
//
// Binary frames are terminal bytes in both directions. Text frames are the
// JSON control frames below. Close vocabulary: 1000 `terminal_ended` is
// authoritative (a complete inventory proved the terminal ended) and 1000
// `detach` is a deliberate client detach; 1011 reasons are retryable.

const ControlFrame = Schema.Union([
  Schema.Struct({ type: Schema.Literal("resize"), cols: Schema.Number, rows: Schema.Number }),
  /** Server keepalive; a client answers with pong. */
  Schema.Struct({ type: Schema.Literal("ping") }),
  Schema.Struct({ type: Schema.Literal("pong") }),
])
export type ControlFrame = typeof ControlFrame.Type

const decodeFrame = Schema.decodeUnknownOption(ControlFrame)

/** Decode one text frame; unknown or malformed frames are undefined, never fatal. */
export const decodeControlFrame = (text: string): ControlFrame | undefined => {
  try {
    return Option.getOrUndefined(decodeFrame(JSON.parse(text)))
  } catch {
    return undefined
  }
}

export const encodeControlFrame = (frame: ControlFrame): string => JSON.stringify(frame)

/**
 * The one dimension rule for resize frames and `cols`/`rows` query params:
 * missing or malformed values take the fallback, everything else is floored
 * and clamped into [1, 1000].
 */
export const clampDimension = (raw: string | number | undefined, fallback: number): number => {
  const parsed = typeof raw === "number" ? raw : Number.parseInt(raw ?? "", 10)
  return Number.isNaN(parsed) ? fallback : Math.max(1, Math.min(1_000, Math.floor(parsed)))
}

/** Close (1000) reason proving the terminal ended — not retryable. */
export const CLOSE_TERMINAL_ENDED = "terminal_ended"
/** Close (1000) reason for a deliberate client detach. */
export const CLOSE_DETACH = "detach"
/** Retryable close (1011) reason: the attach could not be completed. */
export const CLOSE_ATTACH_FAILED = "attach_failed"
/** Retryable close (1011) reason: the multiplexer could not be consulted. */
export const CLOSE_ZMX_UNAVAILABLE = "zmx_unavailable"
/** Retryable close (1011) reason: the client stopped answering pings. */
export const CLOSE_PING_TIMEOUT = "ping_timeout"

/** The contract's `attachTerminal` route as a WebSocket URL. */
export const attachUrl = (baseUrl: URL, terminalId: string, size?: SessionSize): URL => {
  const url = new URL(`/api/v1/terminals/${encodeURIComponent(terminalId)}/attach`, baseUrl)
  url.protocol = baseUrl.protocol === "https:" ? "wss:" : "ws:"
  if (size !== undefined) {
    url.searchParams.set("cols", String(size.cols))
    url.searchParams.set("rows", String(size.rows))
  }
  return url
}

File: packages/ATCKit/Sources/ATCAppServerTransport/AttachProtocol.swift
// AttachProtocol: the pure, transport-free rules of the App Server's
// terminal-attach WebSocket — URL construction, dimension clamping, control
// frames, and close classification.
//
// The wire contract is owned by `packages/attach-protocol/README.md` and the
// server module `app-server/src/terminals/attachProtocol.ts`; the shared JSON
// vectors in `packages/attach-protocol/fixtures/` pin this implementation in
// tests so drift between clients fails a test, not a terminal window.

import Foundation

public enum AttachProtocol {
    /// Server defaults when no size accompanies an attach.
    public static let fallbackCols = 80
    public static let fallbackRows = 24

    /// `ws(s)://<host>/api/v1/terminals/{terminalId}/attach?cols=&rows=` —
    /// scheme derived from the HTTP base URL.
    public static func attachURL(
        base: URL,
        terminalID: String,
        size: (cols: Int, rows: Int)? = nil
    ) -> URL {
        var components = URLComponents(url: base, resolvingAgainstBaseURL: false) ?? URLComponents()
        components.scheme = components.scheme == "https" ? "wss" : "ws"
        // Absolute path, like the server authority's `new URL(path, base)`:
        // a trailing slash on a user-entered base URL must not double up.
        components.path = "/api/v1/terminals/\(terminalID)/attach"
        if let size {
            components.queryItems = [
                URLQueryItem(name: "cols", value: String(size.cols)),
                URLQueryItem(name: "rows", value: String(size.rows)),
            ]
        }
        // The contract's URLs (loopback origin + path segments) always
        // compose into a valid URL.
        return components.url!
    }

    /// The one dimension rule for cols/rows everywhere they occur: floored
    /// and clamped into [1, 1000]; anything non-finite takes the fallback
    /// (matching the server's `Number.isFinite` guard).
    public static func clampDimension(_ value: Double?, fallback: Int) -> Int {
        guard let value, value.isFinite else { return fallback }
        return Int(max(1, min(1000, value.rounded(.down))))
    }

    /// String variant for query parameters: malformed unless the entire
    /// string is numeric (a numeric prefix does not count).
    public static func clampDimension(_ raw: String?, fallback: Int) -> Int {
        clampDimension(raw.flatMap(Double.init), fallback: fallback)
    }

    /// Text-frame control vocabulary. Binary frames are raw terminal bytes
    /// and never appear here.
    public enum ControlFrame: Sendable, Equatable {
        case resize(cols: Int, rows: Int)
        case ping
        case pong

        public var wire: String {
            switch self {
            case .resize(let cols, let rows):
                return #"{"type":"resize","cols":\#(cols),"rows":\#(rows)}"#
            case .ping:
                return #"{"type":"ping"}"#
            case .pong:
                return #"{"type":"pong"}"#
            }
        }

        /// Returns nil for unknown or malformed frames — the protocol says
        /// they are ignored, never fatal. JSONDecoder gives strict number
        /// typing (booleans and numeric strings are malformed, as on the
        /// server), and `clampDimension` makes the Int conversion total, so
        /// no wire value can trap.
        public static func decode(_ text: String) -> ControlFrame? {
            guard let wire = try? JSONDecoder().decode(WireFrame.self, from: Data(text.utf8))
            else { return nil }
            switch wire.type {
            case "resize":
                guard let cols = wire.cols, let rows = wire.rows else { return nil }
                return .resize(
                    cols: clampDimension(cols, fallback: fallbackCols),
                    rows: clampDimension(rows, fallback: fallbackRows)
                )
            case "ping":
                return .ping
            case "pong":
                return .pong
            default:
                return nil
            }
        }

        private struct WireFrame: Decodable {
            let type: String
            let cols: Double?
            let rows: Double?
        }
    }

    /// Reason string the client sends on deliberate detach (close code 1000).
    public static let detachReason = "detach"

    /// What a close tells the client about the terminal.
    public enum CloseDisposition: Sendable, Equatable {
        /// `1000 terminal_ended` — authoritative: a complete multiplexer
        /// inventory proved the session gone; the tombstone is written.
        case terminalEnded
        /// `1000 detach` — deliberate client detach; the session runs on.
        case detach
        /// Everything else (`1011 attach_failed|zmx_unavailable|ping_timeout`,
        /// `1006`, unknown codes): says nothing about the terminal — refetch
        /// it and reattach if still live.
        case retryable
    }

    public static func classifyClose(code: Int, reason: String?) -> CloseDisposition {
        if code == 1000 && reason == "terminal_ended" { return .terminalEnded }
        if code == 1000 && reason == detachReason { return .detach }
        return .retryable
    }
}
